crates/compilers/src/preprocessor.rs

```diff
@@ -0,0 +1,643 @@
+///   - Internal functions
+///   - External functions bodies
+///
+/// Preserves all libraries and interfaces.
+pub(crate) fn interface_representation(
+    content: &str,
+    file: &PathBuf,
+) -> Result<String, EmittedDiagnostics> {
+    let mut spans_to_remove: Vec<Span> = Vec::new();
+    let sess =
+        solar_parse::interface::Session::builder().with_buffer_emitter(Default::default()).build();
+    sess.enter(|| {
+        let arena = solar_parse::ast::Arena::new();
+        let filename = solar_parse::interface::source_map::FileName::Real(file.to_path_buf());
+        let Ok(mut parser) =
+            solar_parse::Parser::from_source_code(&sess, &arena, filename, content.to_string())
+        else {
+            return;
+        };
+        let Ok(ast) = parser.parse_file().map_err(|e| e.emit()) else { return };
+        for item in ast.items {
+            if let solar_parse::ast::ItemKind::Contract(contract) = &item.kind {
+                if contract.kind.is_interface() | contract.kind.is_library() {
+                    continue;
+                }
+                for contract_item in contract.body.iter() {
+                    if let solar_parse::ast::ItemKind::Function(function) = &contract_item.kind {
+                        let is_exposed = match function.kind {
+                            // Function with external or public visibility
+                            solar_parse::ast::FunctionKind::Function => {
+                                function.header.visibility.is_some_and(|visibility| {
+                                    visibility == Visibility::External
+                                        || visibility == Visibility::Public
+                                })
+                            }
+                            solar_parse::ast::FunctionKind::Constructor
+                            | solar_parse::ast::FunctionKind::Fallback
+                            | solar_parse::ast::FunctionKind::Receive => true,
+                            // Other (modifiers)
+                            _ => false,
+                        };
+
+                        // If function is not exposed we remove the entire span (signature and
+                        // body). Otherwise we keep function signature and
+                        // remove only the body.
+                        if !is_exposed {
+                            spans_to_remove.push(contract_item.span);
+                        } else {
+                            spans_to_remove.push(function.body_span);
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    // Return if any diagnostics emitted during content parsing.
+    if let Err(err) = sess.emitted_errors().unwrap() {
+        let e = err.to_string();
+        trace!("failed parsing {file:?}: {e}");
+        return Err(err);
+    }
+
+    let mut content = content.to_string();
+    let mut offset = 0;
+
+    for span in spans_to_remove {
```

Comment:
we should have a module that handles `(source: &str, updates: impl Iterator<Item = (&str, &str)>) -> String` since we are re-implementing this logic in multiple places

Reply:
I put it in compilers lib with https://github.com/foundry-rs/compilers/pull/252/commits/594c09b0a111f5af3c43cadbe5dad0bb05545c71 lmk if this works or you see a better place to move